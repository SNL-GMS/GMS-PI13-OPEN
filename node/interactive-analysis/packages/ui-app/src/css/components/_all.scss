@import 'about-menu';
@import 'body';
@import 'list';
@import 'login';
@import 'quadra-grid';
@import 'workspace';
