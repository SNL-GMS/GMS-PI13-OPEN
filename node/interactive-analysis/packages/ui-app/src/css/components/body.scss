body.bp3-dark {
  background-color: $gms-background;
}

.bp3-navbar {
  padding: 0;
}

.bp3-control input:focus ~ .bp3-control-indicator {
  outline: none;
  outline-offset: 0px;
}
.bp3-control {
  margin-bottom: 0px;
}
.bp3-control:not(.checkbox-horizontal) + .bp3-control:not(.checkbox-horizontal) {
  margin-top: 10px;
}
.bp3-control.checkbox-horizontal + .bp3-control.checkbox-horizontal {
  margin-left: 8px;
}
.bp3-dark .bp3-control.bp3-switch .bp3-control-indicator {
  background: $gms-popover-background;
  margin-left: 4px;
}
.bp3-dark .bp3-control.bp3-switch .bp3-control-indicator::before {
  background: $gms-main;
}
.bp3-dark .bp3-control.bp3-switch:hover .bp3-control-indicator {
  background: $gms-soft;
}

.bp3-dark .bp3-button:not([class*='bp3-intent-']) {
  background: $gms-button;
}
.bp3-dark .bp3-button:not([class*='bp3-intent-']):hover {
  background: $gms-button-hover;
}
.bp3-dark .bp3-button.bp3-minimal {
  background: none;
  background-color: none;
}

.bp3-dark .bp3-popover .bp3-popover-content,
.bp3-popover.bp3-dark .bp3-popover-content {
  background-color: $gms-popover-background;
  color: $gms-prominent;
}

.bp3-dialog {
  padding-bottom: 0.5rem;
}
.bp3-dark .bp3-dialog {
  background-color: $gms-dialog-background;
}

.bp3-dialog-header {
  box-shadow: none;
}

.bp3-dark .bp3-dialog-header {
  box-shadow: none;
}
*::-webkit-scrollbar {
  width: 10px;
  height: 10px;
}

*::-webkit-scrollbar-track {
  background-color: $gms-scrollbar-background;
  width: 10px;
}

*::-webkit-scrollbar-thumb {
  background: $gms-scrollbar;
  border-radius: 2px;
}

:focus {
  outline: none;
  outline-offset: 0px;
}

// override some focus styles from blueprint that dont look good.
.bp3-control input:focus ~ .bp3-control-indicator {
  outline: none;
  outline-offset: 0px;
}

.max {
  position: absolute;
  top: 0px;
  right: 0px;
  bottom: 0px;
  left: 0px;
}

.duo-grid-left {
  position: absolute;
  top: 22%;
  right: 0%;
  left: 1.5%;
  width: 48%;
  height: 18%;
}

.duo-grid-right {
  position: absolute;
  top: 22%;
  right: 0%;
  left: 52%;
  width: 48%;
  height: 18%;
}

.bp3-menu {
  min-width: auto !important;
}

.gms-disconnected {
  position: absolute;
  top: 0px;
  right: 0px;
  bottom: 0px;
  left: 0px;
  z-index: 100;
  background-color: rgba($gms-background, 0.85);
  display: flex;
  justify-content: center;
  align-items: center;
}
