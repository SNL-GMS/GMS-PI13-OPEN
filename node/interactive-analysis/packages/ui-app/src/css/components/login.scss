.login-screen-body {
  background-color: $gms-prominent-background;
  height: 100%;
}

.login-container {
  font-size: 16px;
  width: 100%;
  line-height: 32px;
  display: flex;
  align-items: center;
  height: 100%;
  flex-direction: column;
  justify-content: center;
}

.login-row {
  display: flex;
  width: 380px;
  align-items: flex-start;
  margin-top: 12px;
}

.login-row.login-button {
  justify-content: flex-end;
}

.login-label {
  margin-right: 16px;
  width: 84px;
}

.login-input {
  border-radius: 0;
  color: $analyst-requires-review;
  background-color: $gms-prominent-background;
  font-size: 16px;
  width: 280px;
  line-height: 16px;
  margin-top: 3px;
  border: none;
  border-bottom: 1px solid $gms-soft;
}

.login-legal {
  max-width: 660px;
  line-height: 12px;
  font-size: 12px;
  text-align: justify;
}

.app-menu-button {
  padding-right: 1.25em;
  padding-left: 1em;
  border-left: 1px solid $gms-prominent;
  border-radius: 0px;
}
