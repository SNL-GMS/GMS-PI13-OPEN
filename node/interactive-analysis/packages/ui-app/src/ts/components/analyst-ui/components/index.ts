export { ReduxApolloWaveformDisplay as WaveformDisplay } from './waveform-display';
export { Events } from './events';
export { History } from './history';
export { SignalDetections } from './signal-detections';
export { Workflow } from './workflow';
export { ReduxApolloMap as Map } from './map';
export { Location } from './location';
export { Magnitude } from './magnitude';

export { ReduxApolloAzimuthSlowness as AzimuthSlowness } from './azimuth-slowness/azimuth-slowness-container';
