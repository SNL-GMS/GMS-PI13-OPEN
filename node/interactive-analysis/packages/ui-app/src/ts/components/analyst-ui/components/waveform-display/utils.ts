import { EventTypes } from '@gms/common-graphql';
import { isSdInstantMeasurementValue } from '~analyst-ui/common/utils/instance-of-util';

export interface Offset {
  channelId: string;
  offset: number;
}

/**
 * Calculate offsets based on station with earliest arrival.
 */
export const calculateOffsets = (
  featurePredictions: EventTypes.FeaturePrediction[],
  baseChannelId: string,
  phaseToOffset: string
): Offset[] => {
  if (!featurePredictions || featurePredictions.length <= 0) {
    return [];
  }
  const filteredPredictions: EventTypes.FeaturePrediction[] = featurePredictions.filter(
    fp => fp.phase === phaseToOffset
  );

  // If no FPs were found for the phase selected, return nothing
  if (filteredPredictions.length <= 0) {
    return [];
  }

  filteredPredictions.sort((a, b) => {
    if (
      isSdInstantMeasurementValue(a.predictedValue) &&
      isSdInstantMeasurementValue(b.predictedValue)
    ) {
      const aValue = a.predictedValue.value;
      const bValue = b.predictedValue.value;
      return aValue - bValue;
    }
  });

  const offsets: Offset[] = [];
  const baseChannelFeaturePrediction = filteredPredictions.find(
    fp => fp.channelName === baseChannelId
  );

  if (
    baseChannelFeaturePrediction &&
    isSdInstantMeasurementValue(baseChannelFeaturePrediction.predictedValue)
  ) {
    const baseTime: number = baseChannelFeaturePrediction.predictedValue.value;

    filteredPredictions.forEach(sortedFeaturePrediction => {
      if (isSdInstantMeasurementValue(sortedFeaturePrediction.predictedValue)) {
        const offset: Offset = {
          channelId: sortedFeaturePrediction.channelName,
          offset: baseTime - sortedFeaturePrediction.predictedValue.value
        };
        offsets.push(offset);
      }
    });
  }
  return offsets;
};
