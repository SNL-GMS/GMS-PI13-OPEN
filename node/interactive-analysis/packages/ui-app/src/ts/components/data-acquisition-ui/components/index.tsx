export { StatusConfiguration } from './status-configuration';
export { StationConfiguration } from './station-configuration';
export { StationInformation } from './station-information';
export { TransferGaps } from './transfer-gaps';
export { ConfigureStationGroups } from './configure-station-groups';
export { SohOverview } from './soh-overview';
export { StationStatistics } from './station-statistics';
export { SohEnvironment } from './soh-environment';
export { SohEnvironmentHistory } from './environment-history';
export { SohLagHistory } from './soh-lag-history';
export { SohMissingHistory } from './soh-missing-history';
export { SohLag, SohMissing, SohTimeliness } from '~data-acquisition-ui/components/soh-bar-chart';
