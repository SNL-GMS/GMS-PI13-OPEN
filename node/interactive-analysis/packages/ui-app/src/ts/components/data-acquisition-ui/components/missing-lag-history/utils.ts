import { SohTypes } from '@gms/common-graphql';
import {
  determinePrecisionByType,
  MILLISECONDS_IN_SECOND,
  setDecimalPrecisionAsNumber,
  toDateString,
  toTimeString
} from '@gms/common-util';
import { ValueType } from '@gms/common-util/lib/types/value-type';
import { ChartTypes } from '@gms/ui-core-components';
import { Line } from '@gms/ui-core-components/lib/components/charts/types';
import { DistinctColorPalette } from '@gms/ui-util';
import { dataAcquisitionUIConfig } from '~components/data-acquisition-ui/config';
import { HistoricalSohQueryData } from '~components/data-acquisition-ui/react-apollo-components/historical-soh-query';
import { BarLineChartData } from '~components/data-acquisition-ui/shared/chart/types';
import { THOUSAND } from './constants';
import { MISSING_LAG } from './types';

/** Returns based on the type provided */
export const getName = (type: MISSING_LAG): string =>
  type === SohTypes.SohMonitorType.LAG ? 'Lag' : 'Missing';

/** Returns the label based on the type provided */
export const getLabel = (type: MISSING_LAG): string => `${getName(type)} data`;

/** Returns the data type of the data based on the type provided */
export const getDataType = (type: MISSING_LAG): string =>
  type === SohTypes.SohMonitorType.LAG ? 's' : '%';

/** A helper function for formatting chart tic labels; formats numbers to represent thousandths */
export const chartTickFormatForThousands = (value: number) =>
  value >= THOUSAND ? `${value / THOUSAND}k` : value;

/** Returns the bar chart Y axis label based on the type provided */
export const barChartYAxisLabel = (type: MISSING_LAG): string =>
  `Average ${getName(type)} (${getDataType(type)})`;

/** Returns the bar chart X axis label */
export const barChartXAxisLabel = (): string => 'Channel Name';

/** The bar chart Y axis tic formatter */
export const barChartYAxisTicFormat = chartTickFormatForThousands;

/** The bar chart X axis formatter */
export const barChartXAxisTicFormat = (stationName: string) => (name: string) =>
  name && name.replace ? name.replace(`${stationName}.`, '') : name;

/** Returns the line chart Y axis label based on the type provided */
export const lineChartYAxisLabel = (type: MISSING_LAG): string =>
  `${getName(type)} (${getDataType(type)})`;

/** Returns the line chart X axis label */
export const lineChartXAxisLabel = (): string => 'Time';

/** The line chart Y axis tic formatter */
export const lineChartYAxisTicFormat = chartTickFormatForThousands;

/** The line chart X axis tic formatter */
export const lineChartXAxisTicFormat = (timestamp: number) => {
  const dateString = toDateString(timestamp / MILLISECONDS_IN_SECOND);
  const timeString = toTimeString(timestamp / MILLISECONDS_IN_SECOND);
  return `${dateString}\n${timeString}`;
};

export const getChartHeight = (targetHeightPx: number) =>
  Math.max(targetHeightPx, dataAcquisitionUIConfig.dataAcquisitionUserPreferences.minChartHeightPx);

/**
 * Helper function to return properly scaled value from MonitorValue value
 */
const getMonitorValue = (value: number, monitorType: SohTypes.SohMonitorType) =>
  monitorType === SohTypes.SohMonitorType.LAG
    ? setDecimalPrecisionAsNumber(value) / MILLISECONDS_IN_SECOND
    : setDecimalPrecisionAsNumber(value);

/**
 * Iterate over a channel's values and remove duplicate values used by the LineDef
 * @param values The values to iterate over
 * @param monitorType SohMonitor type i.e. lag or Missing
 * @param calculationTimes The y value in the line plot (time)
 *
 * @return The line (x,y values), with duplicate x values dropped and
 *         the average of all the values is used by the Bar chart in the BarDefinition.
 */
const lineValuesAndAverage = (
  values: number[],
  monitorType: SohTypes.SohMonitorType,
  calculationTimes: number[]
): [Line, number] => {
  // Total value of all the MonitorValue values used in calculating the average value for bar chart
  let sumValue = 0;
  const line: Line = values
    .map((value, index: number) => {
      const currentValue = getMonitorValue(value, monitorType);
      // Add current value
      sumValue += currentValue;

      // If this is not the first value or last value look to see if it can be dropped
      if (index > 0 && index < values.length - 1) {
        // Get the value before and the value after current to compare to
        const beforeValue = getMonitorValue(values[index - 1], monitorType);
        const afterValue = getMonitorValue(values[index + 1], monitorType);
        if (currentValue === beforeValue && currentValue === afterValue) {
          return undefined;
        }
      }
      return {
        x: calculationTimes[index],
        y: currentValue
      };
    })
    .filter(value => value !== undefined);

  // Return the points of the line and the average
  // Average: Total sum of the values and divide by the number of points
  return [line, sumValue / values.length];
};

/**
 * Returns the chart data needed to populate the bar and line charts for
 * the historical data.
 * @param props the missing/lag history props
 * @param context the historical soh query data
 */
export const getChartData = (
  monitorType: MISSING_LAG,
  station: SohTypes.UiStationSoh,
  context: HistoricalSohQueryData,
  colorPalette: DistinctColorPalette,
  valueType: ValueType
): BarLineChartData => {
  const monitorValues = context.data.monitorValues;

  const categories: { x: string[]; y: string[] } = {
    x: monitorValues.map(monitor => monitor.channelName),
    y: []
  };

  const lineDefs: ChartTypes.LineDefinition[] = [];
  const barDefs: ChartTypes.BarDefinition[] = [];

  // Iterate through each channel monitor type building the line definition
  // and bar definition and add them to their respective arrays
  monitorValues.forEach(monitorValue => {
    // Get the Point (x,y) values used in the line definition
    // and average value value of x points used in the bar definition
    const [lineValues, avgValue] = lineValuesAndAverage(
      monitorValue.valuesByType[monitorType].values,
      monitorType,
      context.data.calculationTimes
    );

    // Add the line definition for this channel
    lineDefs.push({
      id: monitorValue.channelName,
      color: colorPalette.getColorString(monitorValue.channelName),
      value: lineValues
    });
    // Add the bar definition for this channel
    barDefs.push({
      id: monitorValue.channelName,
      color: colorPalette.getColorString(monitorValue.channelName),
      value: {
        x: categories.x.find(name => name === monitorValue.channelName),
        y: determinePrecisionByType(avgValue, valueType, false) as number
      }
    });
  });

  // Valid the bar definitions
  barDefs.forEach(
    (barDef, index) =>
      barDef.id !== categories.x[index] &&
      // tslint:disable-next-line: no-console
      console.error('channel names of historical data may not match actual values')
  );

  const statuses = station.channelSohs.map(c =>
    c.allSohMonitorValueAndStatuses.find(v => v.monitorType === monitorType)
  );

  const thresholdsMarginal = statuses.map(s => s.thresholdMarginal);
  const thresholdsBad = statuses.map(s => s.thresholdBad);

  return { categories, lineDefs, barDefs, thresholdsMarginal, thresholdsBad };
};
