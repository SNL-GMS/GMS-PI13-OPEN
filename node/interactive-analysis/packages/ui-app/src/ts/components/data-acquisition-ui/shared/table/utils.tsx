import { SohTypes } from '@gms/common-graphql';
import { SohStatusSummary } from '@gms/common-graphql/lib/graphql/soh/types';
import { MILLISECONDS_IN_SECOND, setDecimalPrecision } from '@gms/common-util';
import { uniq } from 'lodash';
import React from 'react';
import { dataAcquisitionUserPreferences } from '~components/data-acquisition-ui/config';
import { messageConfig } from '~components/data-acquisition-ui/config/message-config';
import {
  DisabledStationSohContextMenu,
  StationSohContextMenu
} from '../context-menus/stations-cell-context-menu';
import { CellData } from './types';

const nonContributingTooltipMessage =
  messageConfig.tooltipMessages.stationStatistics.nonContributingCell;
const nullTooltipMessage = messageConfig.tooltipMessages.stationStatistics.nullCell;
const notReceivedTooltipMessage = messageConfig.tooltipMessages.stationStatistics.notReceivedCell;

/**
 * Returns the height of a row, based on the user preferences, plus a border.
 * This helps get around a linter bug that doesn't see types for values in preferences
 */
export const getRowHeightWithBorder: () => number = () => {
  const defaultBorderSize = 4;
  const rowHeight: number = dataAcquisitionUserPreferences.tableRowHeightPx;
  return rowHeight + defaultBorderSize;
};

/**
 * Returns the height of a row, based on the user preferences, plus a border.
 * This helps get around a linter bug that doesn't see types for values in preferences
 */
export const getHeaderHeight: () => number = () => {
  const extraHeight = 4;
  const rowHeight: number = getRowHeightWithBorder();
  return rowHeight + extraHeight;
};

export const formatSohValue = (value: number): string => {
  if (isNaN(value) || value === null || value === undefined) {
    return 'Unknown';
  }
  return setDecimalPrecision(value);
};

export const sharedSohTableClasses = `
  with-separated-rows
  ag-theme-dark
  ag-theme-dark--soh
  `;

/**
 * Checks if data is null or undefined
 * @param data data to check
 */
export const isNullData = (data: CellData) => data?.value === null || data?.value === undefined;

/**
 * Checks if status is null or undefined
 * @param data CellData to check
 */
export const isNullStatus = (data: CellData) => data?.status === null || data?.status === undefined;

export enum CellStatus {
  GOOD = 'good',
  MARGINAL = 'marginal',
  BAD = 'bad',
  NON_CONTRIBUTING = 'non-contributing'
}

export enum DataReceivedStatus {
  RECEIVED = 'received',
  NOT_ENOUGH_DATA = 'not-enough-data',
  NOT_RECEIVED = 'not-received'
}

const sohStatusMatchesCellStatus = (statusSummary: SohTypes.SohStatusSummary) =>
  Object.values<string>(CellStatus).includes(statusSummary?.toLowerCase());

/**
 * Return a CellStatus that determines the cell status, or non-contributing
 * @param data the cell data to check
 */
export const getCellStatus = (
  status: SohTypes.SohStatusSummary,
  isContributing: boolean = true
) => {
  if (isContributing) {
    if (sohStatusMatchesCellStatus(status)) {
      return (status.toLowerCase() as unknown) as CellStatus;
    }
  }
  return CellStatus.NON_CONTRIBUTING;
};

/**
 * Interprets the cell data to determine if there were any data problems, and of what type.
 * If data is a number in the case of station stats, check if number is defined.
 * @param data the cell data/number to check
 */
export const getDataReceivedStatus = (data: CellData | number) => {
  if (typeof data === 'number') {
    return data !== undefined ? DataReceivedStatus.RECEIVED : DataReceivedStatus.NOT_ENOUGH_DATA;
  }
  return isNullStatus(data)
    ? DataReceivedStatus.NOT_RECEIVED
    : isNullData(data)
    ? DataReceivedStatus.NOT_ENOUGH_DATA
    : DataReceivedStatus.RECEIVED;
};

/**
 * RECEIVED if any data was received.
 * NOT_ENOUGH_DATA if none were received and at least one has NOT_ENOUGH_DATA.
 * NOT_RECEIVED if none were ever received.
 * @param data a cell to check
 */
export const getDataReceivedStatusRollup = (data: CellData[]) =>
  data.reduce<DataReceivedStatus>((worstStatus: DataReceivedStatus, cell) => {
    const cellStatus = getDataReceivedStatus(cell);
    if (cellStatus === DataReceivedStatus.RECEIVED || worstStatus === DataReceivedStatus.RECEIVED) {
      return DataReceivedStatus.RECEIVED;
    }
    if (
      cellStatus === DataReceivedStatus.NOT_ENOUGH_DATA ||
      worstStatus === DataReceivedStatus.NOT_ENOUGH_DATA
    ) {
      return DataReceivedStatus.NOT_ENOUGH_DATA;
    }
    return DataReceivedStatus.NOT_RECEIVED;
  }, DataReceivedStatus.NOT_RECEIVED);

/**
 * Sets the tooltip based on contributing, null status or null value
 * @param data CellData to check
 */
export const setTooltip = (data: CellData) =>
  isNullStatus(data)
    ? notReceivedTooltipMessage
    : !data.isContributing
    ? nonContributingTooltipMessage
    : isNullData(data)
    ? nullTooltipMessage
    : '';

/**
 * Takes in a times to test and determines the delta time to compare against sohStationStaleTimeMS.
 * returns true if calculated time is greater than sohStationStaleTimeMS
 *
 * @param sohStationStaleTimeMS time limit for stale time to test against
 * @param timeToTest number in seconds gets converted to milliseconds
 */
export const isSohStationStaleTimeMS = (
  timeToTest: number,
  sohStationStaleTimeMS: number
): boolean => {
  // TODO: bring this into a util area
  // get milliseconds for current time
  const currentTime: number = new Date().valueOf();
  // get milliseconds for the time we are testing
  const timeToTestMS: number = timeToTest * MILLISECONDS_IN_SECOND;
  const deltaTime: number = Math.abs(timeToTestMS - currentTime);
  // send back true if we are past the configured time
  return deltaTime > sohStationStaleTimeMS;
};
/**
 * Helper function to determine if any of the selected Station SOH are stale or
 * are already acknowledged
 * @param stationNames selected
 * @returns boolean
 */
export const isAcknowledgeEnabled = (
  stationNames: string[],
  stationSohs: SohTypes.UiStationSoh[],
  sohStationStaleTimeMS: number
) => {
  let isDisabled = false;
  uniq(stationNames).forEach((name: string) => {
    const soh = stationSohs.find(entry => entry.stationName === name);
    if (soh && soh.time) {
      isDisabled = isSohStationStaleTimeMS(soh.time, sohStationStaleTimeMS)
        ? true
        : !soh.needsAcknowledgement || isDisabled;
    } else {
      isDisabled = true;
    }
  });
  return !isDisabled;
};

/**
 * Method returns the appropriate Station SOH Context Menu
 * @param stationNames the selected station names
 * @param stationSohs the station soh data
 * @param sohStationStaleTimeMS the station stale timeout in milliseconds
 * @param acknowledgeCallback the callback for acknowledging
 * @return StationSohContextMenu
 */
export const acknowledgeContextMenu = (
  stationNames: string[],
  stationSohs: SohTypes.UiStationSoh[],
  sohStationStaleTimeMS: number,
  acknowledgeCallback: (stationIds: string[], comment?: string) => void
) => {
  const sohContextMenuProps = {
    stationNames,
    // tslint:disable-next-line: no-unbound-method
    acknowledgeCallback
  };
  return isAcknowledgeEnabled(stationNames, stationSohs, sohStationStaleTimeMS) ? (
    <StationSohContextMenu {...sohContextMenuProps} />
  ) : (
    <DisabledStationSohContextMenu {...sohContextMenuProps} />
  );
};

/**
 * Custom comparator for comparing cell values.
 * Handles undefined
 * @param a the first value
 * @param b the second value
 * @returns number indicating sort order (standard comparator return)
 */
export const compareCellValues = (a: number, b: number) => {
  if (a === undefined && b === undefined) {
    return 0;
  }
  if (a === undefined) {
    return -1;
  }
  if (b === undefined) {
    return 1;
  }
  return a - b;
};

/**
 * Compares statuses
 * @param a first soh status
 * @param b second soh status
 * @returns the worst status
 */
export const getWorseStatus = (a: SohStatusSummary, b: SohStatusSummary): SohStatusSummary => {
  if (a === SohStatusSummary.BAD || b === SohStatusSummary.BAD) {
    return SohStatusSummary.BAD;
  }
  if (a === SohStatusSummary.MARGINAL || b === SohStatusSummary.MARGINAL) {
    return SohStatusSummary.MARGINAL;
  }
  if (a === SohStatusSummary.GOOD || b === SohStatusSummary.GOOD) {
    return SohStatusSummary.GOOD;
  }
  return SohStatusSummary.NONE;
};

export const getWorstCapabilityRollup = (groups: SohTypes.StationSohCapabilityStatus[]) =>
  groups &&
  groups.length &&
  groups.reduce(
    (
      worstFound: SohTypes.SohStatusSummary,
      capabilityStatus: SohTypes.StationSohCapabilityStatus
    ) => getWorseStatus(worstFound, capabilityStatus?.sohStationCapability),
    SohTypes.SohStatusSummary.NONE
  );
