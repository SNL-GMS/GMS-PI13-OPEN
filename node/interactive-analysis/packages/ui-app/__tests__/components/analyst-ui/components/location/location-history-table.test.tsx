import React from 'react';
import { EventUtils } from '../../../../../src/ts/components/analyst-ui/common/utils';
import { LocationHistory } from '../../../../../src/ts/components/analyst-ui/components/location/components/location-history';
import * as LocationData from '../../../../__data__/location-data';

// tslint:disable-next-line:no-implicit-dependencies no-var-requires no-require-imports
const Enzyme = require('enzyme');
// set up window alert and open so we don't see errors
window.alert = jest.fn();
window.open = jest.fn();

it('Location History Table Renders a single location entry', () => {
  const wrapper = Enzyme.shallow(
    <LocationHistory
      event={LocationData.event}
      location={{
        selectedLocationSolutionSetId: undefined,
        selectedLocationSolutionId: undefined,
        selectedPreferredLocationSolutionSetId: undefined,
        selectedPreferredLocationSolutionId: undefined
      }}
      setSelectedLocationSolution={(locationSolutionSetId: string, locationSolutionId: string) => {
        /* no-op */
      }}
      setSelectedPreferredLocationSolution={(
        preferredLocationSolutionSetId: string,
        preferredLocationSolutionId: string
      ) => {
        /* no-o */
      }}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

// Unit test for checking that the correct, configured preferred restraints is picked
describe('When location is set to save, correct preferred location is set', () => {
  it('The osd event hypothesis should match expected result', () => {
    const preferredId = EventUtils.getPreferredLocationSolutionIdFromEventHypothesis(
      LocationData.eventHypothesisWithLocationSets
    );
    const hardCodedCorrectId = '3a06fac7-46ad-337e-a8da-090a1cc801a1';
    expect(preferredId).toEqual(hardCodedCorrectId);
  });
});
