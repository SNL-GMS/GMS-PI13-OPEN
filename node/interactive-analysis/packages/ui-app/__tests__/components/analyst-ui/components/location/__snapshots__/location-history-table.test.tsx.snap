// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location History Table Renders a single location entry 1`] = `
<div
  className="ag-theme-dark table-container"
>
  <div
    className="list-wrapper"
  >
    <div
      className="max"
    >
      <ClientSideTable
        autoGroupColumnDef={
          Object {
            "cellRendererFramework": [Function],
            "comparator": [Function],
            "field": "save",
            "headerName": "Save",
            "headerTooltip": "Save this location set when the event is saved",
            "width": 65,
          }
        }
        columnDefs={
          Array [
            Object {
              "cellRendererFramework": [Function],
              "field": "locationSetId",
              "headerName": "ASBDUSA",
              "hide": true,
              "sortable": true,
            },
            Object {
              "cellStyle": Object {
                "text-align": "left",
              },
              "field": "locType",
              "headerName": "Type",
              "headerTooltip": "Event Location Type (Standard or Master)",
              "resizable": true,
              "width": 90,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "field": "lat",
              "headerName": "Lat (°)",
              "headerTooltip": "Latitude of event location",
              "resizable": true,
              "valueFormatter": [Function],
              "width": 70,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "field": "lon",
              "headerName": "Lon (°)",
              "headerTooltip": "Longitude of event location",
              "resizable": true,
              "valueFormatter": [Function],
              "width": 70,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "field": "depth",
              "headerName": "Depth (km)",
              "headerTooltip": "Depth of event location",
              "resizable": true,
              "valueFormatter": [Function],
              "width": 90,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "field": "time",
              "headerName": "Time",
              "headerTooltip": "Time of event",
              "resizable": true,
              "width": 210,
            },
            Object {
              "cellStyle": Object {
                "text-align": "left",
              },
              "field": "restraint",
              "headerName": "Restraint",
              "headerTooltip": "Depth restraint of location calculation",
              "resizable": true,
              "width": 140,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "field": "smajax",
              "headerName": "Semi Major (km)",
              "headerTooltip": "Semi-major axis length of error ellipse",
              "resizable": true,
              "valueFormatter": [Function],
              "width": 122,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "field": "sminax",
              "headerName": "Semi Minor (km)",
              "headerTooltip": "Semi-minor axis length of error ellipse",
              "resizable": true,
              "valueFormatter": [Function],
              "width": 122,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "field": "strike",
              "headerName": "Strike (°)",
              "headerTooltip": "Strike of major axis of error ellipse",
              "resizable": true,
              "valueFormatter": [Function],
              "width": 70,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "field": "stdev",
              "headerName": "Std Dev",
              "headerTooltip": "Standard deviation of the observed",
              "resizable": true,
              "valueFormatter": [Function],
              "width": 70,
            },
            Object {
              "cellStyle": Object {
                "text-align": "right",
              },
              "field": "count",
              "headerName": "Count",
              "hide": true,
              "resizable": true,
              "sortable": true,
              "width": 95,
            },
            Object {
              "cellRendererFramework": [Function],
              "cellStyle": Object {
                "display": "flex",
                "justify-content": "center",
              },
              "field": "preferred",
              "headerName": "Preferred",
              "headerTooltip": "Mark an event location as preferred",
              "width": 80,
            },
          ]
        }
        context={Object {}}
        getRowNodeId={[Function]}
        getRowStyle={[Function]}
        gridOptions={
          Object {
            "rowClass": "location-history-row",
          }
        }
        groupMultiAutoColumn={true}
        id="table-location-history"
        key="table-location-history"
        onCellClicked={[Function]}
        onGridReady={[Function]}
        onRowClicked={[Function]}
        rowData={
          Array [
            Object {
              "count": 0,
              "depth": 0,
              "depthRestraintType": "UNRESTRAINED",
              "id": "1",
              "isFirstInLSSet": true,
              "isLastInLSSet": true,
              "isLocationSolutionSetPreferred": false,
              "lat": 44.94862,
              "latestLSSId": "5fa862f1-2e2d-4d87-bc66-00177de18c5e",
              "locType": "Standard",
              "locationSetId": "5fa862f1-2e2d-4d87-bc66-00177de18c5e",
              "locationSolutionId": "3a06fac7-46ad-337e-a8da-090a1cc801a1",
              "lon": -106.38442,
              "preferred": false,
              "restraint": "Unrestrained",
              "selectedLocationSolutionSetId": undefined,
              "setSelectedPreferredLocationSolution": [Function],
              "setToSave": [Function],
              "smajax": 59.7138,
              "sminax": 32.3731,
              "stdev": 0,
              "strike": 137.57,
              "time": "2010/05/20T22:01:30.238000",
            },
          ]
        }
        rowSelection="none"
        suppressContextMenu={true}
      />
    </div>
  </div>
</div>
`;
